### ISSUE TITLE
Error while installing Retriever on Linux when using a proxy

### ISSUE THREAD

**Issue by prabh27** [weecology/retriever#268]
Error is as follows: 
Traceback (most recent call last):
  File "/usr/bin/retriever", line 9, in <module>
    load_entry_point('retriever==1.7.0', 'console_scripts', 'retriever')()
  File "/usr/lib/python2.7/dist-packages/retriever/__main__.py", line 31, in main
    check_for_updates(graphical=False if 'darwin' in platform.platform().lower() else True)
  File "/usr/lib/python2.7/dist-packages/retriever/lib/repository.py", line 74, in check_for_updates
    init.run()
  File "/usr/lib/python2.7/dist-packages/retriever/lib/repository.py", line 143, in run
    version_file = urllib.urlopen(REPOSITORY + "version.txt")
  File "/usr/lib/python2.7/urllib.py", line 87, in urlopen
    return opener.open(url)
  File "/usr/lib/python2.7/urllib.py", line 208, in open
    return getattr(self, name)(url)
  File "/usr/lib/python2.7/urllib.py", line 437, in open_https
    h.endheaders(data)
  File "/usr/lib/python2.7/httplib.py", line 969, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 829, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 791, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 1176, in connect
    self.sock = ssl.wrap_socket(sock, self.key_file, self.cert_file)
  File "/usr/lib/python2.7/ssl.py", line 487, in wrap_socket
    ciphers=ciphers)
  File "/usr/lib/python2.7/ssl.py", line 243, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 405, in do_handshake
    self._sslobj.do_handshake()
IOError: [Errno socket error] [Errno 1] _ssl.c:510: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol

**Comment by ethanwhite** [comment]
Thanks for the bug report @prabh27. I can't replicate this at the moment. Can you give me some more information on your system configuration including OS and Python versions and the version of openssl since that appears to be what's throwing the error?

**Comment by prabh27** [comment]
System Configuration: Linux 14.04
Python Version : 2.7.6
OpenSSL version : OpenSSL 1.0.1f 6 Jan 2014

**Comment by ethanwhite** [comment]
Thanks. This is basically my setup (minus a couple of minor version differences; Ubuntu 14.10, Python 2.7.8) and basically the same as our continuous integration system, both of which are building just fine at the moment.

Can you try opening Python and running

```
import urllib
urllib.urlopen("https://raw.github.com/weecology/retriever/master/version.txt")
```

If you get the same error try

```
import urllib
urllib.urlopen("https://raw.githubusercontent.com/weecology/retriever/master/version.txt")
```

If you still get the same error try opening a non-https page and see what happens.

Thanks.

**Comment by prabh27** [comment]
Unfortunately, the same error.
I tried to open http://www.google.com and it works fine, but the same error occurs with https://www.fb.com.

**Comment by prabh27** [comment]
This error is occuring due to proxy issues. Whenever there is a proxy server, urllib doesn't open https without proxy, and we have to give the proxies along with the url. I added the proxy in repository.py in the lib/ directory, but the problem still occurs. Where else should I change the code so that this issue is solved.
This can be integrated in the main directory as well so that others using proxy don't face the same problem.

**Comment by ethanwhite** [comment]
Nice job tracking down the issue. There are three uses of `urlopen` in the code:

```
ethan@oryx:~/Informatics/Retriever/src (master)$ grep -nr urlopen *
lib/repository.py:104:       version_file = urllib.urlopen(MASTER_BRANCH + "version.txt")
lib/repository.py:143:       version_file = urllib.urlopen(REPOSITORY + "version.txt")
lib/engine.py:370:           file = urllib.urlopen(url) 
```
 
The first two uses will definitely experience this problem since they both access github which is always https. The call in engine accesses a number of different sources and in concept could hit https at some point, so we should go ahead and fix it as well.

**Comment by ethanwhite** [comment]
Adding a link to the relevant docs for reference:
https://docs.python.org/2/library/urllib.html#urllib.urlopen

And updating the title of the issue.

**Comment by prabh27** [comment]
New error : 
Traceback (most recent call last):
  File "/usr/local/bin/retriever", line 9, in <module>
    load_entry_point('retriever==1.7.0', 'console_scripts', 'retriever')()
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/__main__.py", line 32, in main
    check_for_updates(graphical=False if 'darwin' in platform.platform().lower() else True)
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 75, in check_for_updates
    init.run()
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 164, in run
    os.path.join(SCRIPT_WRITE_PATH, script_name))
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 27, in download_from_repository
    urllib.urlretrieve(repo + filepath, newpath, reporthook=reporthook)
  File "/usr/lib/python2.7/urllib.py", line 94, in urlretrieve
    return _urlopener.retrieve(url, filename, reporthook, data)
  File "/usr/lib/python2.7/urllib.py", line 240, in retrieve
    fp = self.open(url, data)
  File "/usr/lib/python2.7/urllib.py", line 208, in open
    return getattr(self, name)(url)
  File "/usr/lib/python2.7/urllib.py", line 437, in open_https
    h.endheaders(data)
  File "/usr/lib/python2.7/httplib.py", line 969, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 829, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 791, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 1176, in connect
    self.sock = ssl.wrap_socket(sock, self.key_file, self.cert_file)
  File "/usr/lib/python2.7/ssl.py", line 487, in wrap_socket
    ciphers=ciphers)
  File "/usr/lib/python2.7/ssl.py", line 243, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 405, in do_handshake
    self._sslobj.do_handshake()
IOError: [Errno socket error] [Errno 1] _ssl.c:510: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol

**Comment by prabh27** [comment]
If I run the command as root, I get this : 

(python:29714): IBUS-WARNING **: The owner of /home/prabhsimran/.config/ibus/bus is not root!
Traceback (most recent call last):
  File "/usr/local/bin/retriever", line 9, in <module>
    load_entry_point('retriever==1.7.0', 'console_scripts', 'retriever')()
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/__main__.py", line 32, in main
    check_for_updates(graphical=False if 'darwin' in platform.platform().lower() else True)
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 75, in check_for_updates
    init.run()
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 164, in run
    os.path.join(SCRIPT_WRITE_PATH, script_name))
  File "/usr/local/lib/python2.7/dist-packages/retriever-1.7.0-py2.7.egg/retriever/lib/repository.py", line 27, in download_from_repository
    urllib.urlretrieve(repo + filepath, newpath, reporthook=reporthook)
  File "/usr/lib/python2.7/urllib.py", line 94, in urlretrieve
    return _urlopener.retrieve(url, filename, reporthook, data)
  File "/usr/lib/python2.7/urllib.py", line 240, in retrieve
    fp = self.open(url, data)
  File "/usr/lib/python2.7/urllib.py", line 208, in open
    return getattr(self, name)(url)
  File "/usr/lib/python2.7/urllib.py", line 437, in open_https
    h.endheaders(data)
  File "/usr/lib/python2.7/httplib.py", line 969, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 829, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 791, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 1172, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python2.7/socket.py", line 571, in create_connection
    raise err
IOError: [Errno socket error] [Errno 111] Connection refused

**Comment by prabh27** [comment]
The problem has been solved, but the person will always have to add the proxy himself. I can have a system call in python to export proxy for http and https. Is there any other better way possible?

**Comment by prabh27** [comment]
I have submiited the pull requests, but Travel CI is failing. Please have a look at it. It works fine on my system.

**Comment by ethanwhite** [comment]
Thanks @prabh27! The Travis failure is described in #269 and doesn't have anything to do with your code. I'll try to take a look at the PRs today and provide some feedback.

**Comment by ethanwhite** [comment]
Fixed in f55af10f1767d39fdba65fb4c17beee526f96748.
